apps/hpc-cdm/src/app/pages/operation-cluster-forms.tsx

```diff
@@ -20,71 +20,73 @@ interface Props {
-                      />
-                    );
-                  }
-                }
-                return (
-                  <>
-                    <ClusterNavigation
-                      operation={operation}
-                      cluster={cluster}
-                      showSettingsButton
-                    />
-                    <C.NotFound
-                      strings={t.get(lang, (s) => s.components.notFound)}
-                    />
-                  </>
-                );
-              }}
+              path=""
+              element={
+                operation.reportingWindows.length > 0 ? (
+                  <Navigate
+                    to={paths.operationClusterFormAssignments({
+                      operationId: operation.id,
+                      clusterId: cluster.id,
+                      windowId: getBestReportingWindow(
+                        operation.reportingWindows
+                      ).id,
+                    })}
+                  />
+                ) : (
+                  <C.ErrorMessage
+                    strings={{
+                      title: 'No reporting windows',
+                      info: "This operation doesn't have any reporting windows associated with it",
+                    }}
+                  />
+                )
+              }
             />
-            <Route>
-              {operation.reportingWindows.length > 0 ? (
-                <Redirect
-                  to={paths.operationClusterFormAssignments({
-                    operationId: operation.id,
-                    clusterId: cluster.id,
-                    windowId: getBestReportingWindow(operation.reportingWindows)
-                      .id,
-                  })}
-                />
-              ) : (
-                <C.ErrorMessage
-                  strings={{
-                    title: 'No reporting windows',
-                    info: "This operation doesn't have any reporting windows associated with it",
-                  }}
-                />
-              )}
-            </Route>
-          </Switch>
+          </Routes>
         </div>
       )}
     </AppContext.Consumer>
   );
 };
 
+interface ValidateParamsProps extends Props {
+  lang: LanguageKey;
+}
+const ValidateParams = (props: ValidateParamsProps) => {
+  const { operation, cluster, lang } = props;
+  const params = useParams();
+  const id = Number(params.windowId);
+  if (!isNaN(id)) {
+    const windows = operation.reportingWindows.filter((w) => w.id === id);
+    if (windows.length === 1) {
+      return (
+        <PageOperationClusterFormAssignments
+          window={windows[0]}
+          {...{ operation, cluster }}
+        />
+      );
+    }
+  }
+  return (
+    <>
+      <ClusterNavigation
+        operation={operation}
+        cluster={cluster}
+        showSettingsButton
+      />
+      <C.NotFound strings={t.get(lang, (s) => s.components.notFound)} />
+    </>
+  );
+};
```

Comment:
Second big complaint regarding how you migrated to React Router v6 is that you have a bunch of patterns like this. We can no longer have functions inside `<Route>` elements where we could do pre-processing and verification of URL parameters, but you have done exact thing in all the places where this pattern occurs:

1. Add local "temporary" (you even named it `TemporalForm` in one place) component and process parameters with `useParams()` inside of it
2. Use this temporary component with `<React element>`, now that all elements should be passed via that prop

My biggest complaint is actually that you recognized this as a pattern and then copy pasted same structure all over the place. You named such construct a "temporal" (probably meant "temporary") component in one place - `TemporalForm`, but then seem to have realized why such construct is needed (to validate URL path params) and named it more descriptive `ValidateParams`. You have repeated the same pattern 5 times without moving this into a separate component!

From reading the documentation, the maintainers seem to want us to use `useParams()` in components directly, that is fine, but you shouldn't have introduced a temporary component. What you should have done, according to documentation is exactly what you did in `apps/hpc-cdm/src/app/pages/operation.tsx`. Just use `useParams()` is component directly, no temporary components needed. That way, instead of component's error being handled outside, component will be responsible for checking its own props.

But, like I said, we need to go one step further and have a component that will do this for us, in a generic manner.